Plugins/Autodiff/include/Acts/Plugins/Autodiff/AutodiffExtensionWrapper.hpp

```diff
@@ -0,0 +1,176 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Utilities/Definitions.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+#include "Acts/Utilities/ParameterDefinitions.hpp"
+
+#include <autodiff/forward.hpp>
+#include <autodiff/forward/eigen.hpp>
+
+namespace Acts {
+
+/// @brief Default RKN4 evaluator for autodiff
+template <template <typename> typename basic_extension_t>
+struct AutodiffExtensionWrapper {
+  /// @brief Default constructor
+  AutodiffExtensionWrapper() = default;
+
+  /// Some typedefs
+  using AutodiffScalar = autodiff::dual;
+
+  using AutodiffVector3 = ActsMatrix<AutodiffScalar, 3, 1>;
+  using AutodiffFreeMatrix = ActsMatrix<AutodiffScalar, 8, 8>;
+  using AutodiffFreeVector = ActsMatrix<AutodiffScalar, 8, 1>;
+
+  /// Instance of the base extension
+  basic_extension_t<double> m_doubleExtension;
```

Comment:
@benjaminhuth Could you comment?

Reply:
You find extension templated with `AutodiffScalar` in line 143. The reason for this is hat you need the double-extension to communicate with the "outer world" (The stepper). You cannot do this computation with autodiff, since then we would need to template the stepper itself. So, from the outside the Wrapper behaves exactly like the underlying Extension with double, just internally it computes the jacobian with autodiff. Another advantage is, that with this structure we don't need to differentiate through the adaptive stepsize estimation in the stepper.
The `basic_extension_t<AutodiffScalar>` is just necessary locally, its not necessary that it is a class member. To work with it, you need the `FakeStepper` etc., defined in the private section of the class.
I hope this clarifies it a bit.

Reply:
Ok, thanks. That makes sense to me.